File: scripts/test-msam-api.ts
#!/usr/bin/env node

/**
 * MSAM REST API Integration Test
 * TypeScript version for pi-agent ecosystem
 */

import { request } from 'http';

const API_URL = 'http://localhost:3001';
const API_BASE = '/v1';

interface Response {
  status: number;
  data: any;
}

function httpRequest(
  method: string,
  endpoint: string,
  data?: any
): Promise<Response> {
  return new Promise((resolve, reject) => {
    const url = new URL(API_URL + API_BASE + endpoint);
    const options: any = {
      hostname: url.hostname,
      port: url.port || 8000,
      path: url.pathname,
      method,
      headers: {
        'Content-Type': 'application/json',
      },
    };

    const req = request(options, (res) => {
      let body = '';
      res.on('data', chunk => body += chunk);
      res.on('end', () => {
        try {
          const responseData = body ? JSON.parse(body) : {};
          resolve({ status: res.statusCode || 500, data: responseData });
        } catch (e) {
          resolve({ status: res.statusCode || 500, data: body });
        }
      });
    });

    req.on('error', reject);

    if (data) {
      req.write(JSON.stringify(data));
    }
    req.end();
  });
}

async function testStore(): Promise<boolean> {
  console.log('📤 Testing POST /store...');
  try {
    const result = await httpRequest('POST', '/store', {
      content: 'User tested MSAM REST API successfully (TypeScript)',
      stream: 'episodic',
      arousal: 0.6,
      valence: 0.7,
      encoding_confidence: 0.9,
    });
    console.log(`  Status: ${result.status}`);
    console.log(`  Response: ${JSON.stringify(result.data, null, 2).substring(0, 200)}...`);
    return result.status === 200;
  } catch (error: any) {
    console.log(`  Error: ${error.message}`);
    return false;
  }
}

async function testQuery(): Promise<boolean> {
  console.log('\n📥 Testing POST /query...');
  try {
    const result = await httpRequest('POST', '/query', {
      query: 'What do you know about MSAM REST API TypeScript?'
    });
    console.log(`  Status: ${result.status}`);
    if (result.status === 200) {
      const data = result.data;
      console.log(`  Confidence Tier: ${data.confidence_tier}`);
      console.log(`  Total Tokens: ${data.total_tokens}`);
      console.log(`  Items Returned: ${data.items_returned}`);
      console.log(`  Latency: ${data.latency_ms}ms`);
      if (data.atoms && data.atoms.length > 0) {
        console.log(`  First Atom: ${data.atoms[0].content.substring(0, 80)}...`);
      }
    }
    return result.status === 200;
  } catch (error: any) {
    console.log(`  Error: ${error.message}`);
    return false;
  }
}

async function testContext(): Promise<boolean> {
  console.log('\n📋 Testing POST /context...');
  try {
    const result = await httpRequest('POST', '/context', {});
    console.log(`  Status: ${result.status}`);
    if (result.status === 200) {
      const data = result.data;
      console.log(`  Total Tokens: ${data.total_tokens}`);
      console.log(`  Method: ${data.method}`);
    }
    return result.status === 200;
  } catch (error: any) {
    console.log(`  Error: ${error.message}`);
    return false;
  }
}

async function testStats(): Promise<boolean> {
  console.log('\n📊 Testing GET /stats...');
  try {
    const result = await httpRequest('GET', '/stats');
    console.log(`  Status: ${result.status}`);
    if (result.status === 200) {
      const data = result.data;
      console.log(`  Total Atoms: ${data.total_atoms}`);
      console.log(`  Active Atoms: ${data.active_atoms}`);
      console.log(`  Est Tokens: ${data.est_active_tokens}`);
    }
    return result.status === 200;
  } catch (error: any) {
    console.log(`  Error: ${error.message}`);
    return false;
  }
}

async function main(): Promise<number> {
  console.log('🧪 MSAM REST API Integration Test (TypeScript)\n');
  console.log(`API URL: ${API_URL}`);
  console.log('='.repeat(50));

  const results: [string, boolean][] = [
    ['POST /store', await testStore()],
    ['POST /query', await testQuery()],
    ['POST /context', await testContext()],
    ['GET /stats', await testStats()],
  ];

  console.log('\n' + '='.repeat(50));
  console.log('📋 Test Results:');
  for (const [endpoint, success] of results) {
    const status = success ? '✅ PASS' : '❌ FAIL';
    console.log(`  ${status} ${endpoint}`);
  }

  const allPass = results.every(([, success]) => success);
  if (allPass) {
    console.log('\n✅ All tests passed!');
  } else {
    console.log('\n❌ Some tests failed');
  }

  return allPass ? 0 : 1;
}

main().catch((error: any) => {
  console.log(`\n❌ ERROR: ${error.message}`);
  if (error.code === 'ECONNREFUSED') {
    console.log('   MSAM server not running');
    console.log('   Start it with: cd ~/msam && python -m msam.server');
  }
  process.exit(1);
});

File: scripts/test_msam_api.py
#!/usr/bin/env python3
"""Test MSAM REST API integration"""

import json
import time
import requests

API_URL = "http://localhost:8000"

def test_store():
    """Test storing atoms via API."""
    print("📤 Testing POST /store...")
    response = requests.post(
        f"{API_URL}/store",
        json={
            "content": "User tested MSAM REST API successfully",
            "stream": "episodic",
            "arousal": 0.6,
            "valence": 0.7,
            "encoding_confidence": 0.9
        },
        timeout=10
    )
    print(f"  Status: {response.status_code}")
    print(f"  Response: {json.dumps(response.json(), indent=2)[:200]}")
    return response.status_code == 200

def test_query():
    """Test querying via API."""
    print("\n📥 Testing POST /query...")
    response = requests.post(
        f"{API_URL}/query",
        json={"query": "What do you know about MSAM REST API?"},
        timeout=10
    )
    print(f"  Status: {response.status_code}")
    if response.status_code == 200:
        data = response.json()
        print(f"  Confidence Tier: {data.get('confidence_tier')}")
        print(f"  Total Tokens: {data.get('total_tokens')}")
        print(f"  Items Returned: {data.get('items_returned')}")
        print(f"  Latency: {data.get('latency_ms')}ms")
        if data.get('atoms'):
            print(f"  First Atom: {data['atoms'][0]['content'][:80]}...")
    return response.status_code == 200

def test_context():
    """Test context via API."""
    print("\n📋 Testing POST /context...")
    response = requests.post(
        f"{API_URL}/context",
        json={},
        timeout=10
    )
    print(f"  Status: {response.status_code}")
    if response.status_code == 200:
        data = response.json()
        print(f"  Total Tokens: {data.get('total_tokens')}")
        print(f"  Method: {data.get('method')}")
    return response.status_code == 200

def test_stats():
    """Test stats via API."""
    print("\n📊 Testing GET /stats...")
    response = requests.get(f"{API_URL}/stats", timeout=10)
    print(f"  Status: {response.status_code}")
    if response.status_code == 200:
        data = response.json()
        print(f"  Total Atoms: {data.get('total_atoms')}")
        print(f"  Active Atoms: {data.get('active_atoms')}")
        print(f"  Est Tokens: {data.get('est_active_tokens')}")
    return response.status_code == 200

def main():
    """Run all API tests."""
    print("🧪 MSAM REST API Integration Test\n")
    print(f"API URL: {API_URL}")
    print("=" * 50)

    results = []
    results.append(("POST /store", test_store()))
    results.append(("POST /query", test_query()))
    results.append(("POST /context", test_context()))
    results.append(("GET /stats", test_stats()))

    print("\n" + "=" * 50)
    print("📋 Test Results:")
    for endpoint, success in results:
        status = "✅ PASS" if success else "❌ FAIL"
        print(f"  {status} {endpoint}")

    all_pass = all(success for _, success in results)
    if all_pass:
        print("\n✅ All tests passed!")
    else:
        print("\n❌ Some tests failed")

    return 0 if all_pass else 1

if __name__ == "__main__":
    try:
        exit(main())
    except requests.exceptions.ConnectionError:
        print("\n❌ ERROR: MSAM server not running")
        print("   Start it with: cd ~/msam && python -m msam.serve")
        exit(1)
    except Exception as e:
        print(f"\n❌ ERROR: {e}")
        exit(1)
